File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/281ce8c4a5924dc6b60bf56d2bb318d5.java
import java.io.*;
import java.util.*;
import java.util.function.*;
import java.util.stream.*;

public class 281ce8c4a5924dc6b60bf56d2bb318d5_nc {
    static class Solver {
        BufferedReader in;
        PrintWriter out;

        Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        void solve() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            int n = Integer.parseInt(st.nextToken());
            int k = Integer.parseInt(st.nextToken());

            List<List<Integer>> arrs = new ArrayList<>();
            long[] sums = new long[n];

            for (int i = 0; i < n; i++) {
                int t = Integer.parseInt(in.readLine());
                List<Integer> arr = new ArrayList<>();
                for (int j = 0; j < t; j++) {
                    arr.add(Integer.parseInt(in.readLine()));
                }
                if (t > k) {
                    arr = arr.subList(0, k);
                }
                arrs.add(arr);
                sums[i] = arr.stream().mapToLong(Integer::longValue).sum();
            }

            long[][] dps = new long[13][k + 1];
            long[] dp = new long[k + 1];

            BiConsumer<Integer, Integer> doCopy = (dpi, k1) -> {
                System.arraycopy(dps[dpi], 0, dps[dpi + 1], 0, k1 + 1);
            };

            BiConsumer<Integer, Integer> addToKnap = (i, dpi) -> {
                long v = sums[i];
                int k1 = arrs.get(i).size();
                for (int j = k; j >= k1; j--) {
                    dps[dpi][j] = Math.max(dps[dpi][j], dps[dpi][j - k1] + v);
                }
            };

            long ans = yCombinator((rec, l, r, dpi) -> {
                long result = 0;
                List<Integer> carr = arrs.get(l);
                long[] dpCurrent = dps[dpi];

                if (l + 1 == r) {
                    result = Math.max(result, dpCurrent[k]);
                    long sum = 0;
                    for (int i = 0; i < Math.min(k, carr.size()); i++) {
                        sum += carr.get(i);
                        result = Math.max(result, dpCurrent[k - i - 1] + sum);
                    }
                } else {
                    int m = (l + r) / 2;
                    result = Math.max(result, doSplit(l, m, m, r, dpi, addToKnap, doCopy, rec));
                    result = Math.max(result, doSplit(m, r, l, m, dpi, addToKnap, doCopy, rec));
                }
                return result;
            }).apply(0, n, 0);

            out.println(ans);
        }

        long doSplit(int l1, int r1, int l2, int r2, int dpi, BiConsumer<Integer, Integer> addToKnap, BiConsumer<Integer, Integer> doCopy, BiFunction<BiFunction<Integer, Integer, Integer>, Integer, Long> rec) {
            doCopy.accept(dpi, k);
            for (int i = l1; i < r1; i++) {
                addToKnap.accept(i, dpi + 1);
            }
            return rec.apply(this::doSplit, l2, r2, dpi + 1);
        }

        <R> Function<Integer, R> yCombinator(BiFunction<BiFunction<Integer, Integer, Integer>, Integer, R> f) {
            return new Function<Integer, R>() {
                @Override
                public R apply(Integer arg) {
                    return f.apply(this::apply, arg);
                }
            };
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        new Solver(in, out).solve();
        out.close();
    }
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/281ce8c4a5924dc6b60bf56d2bb318d5.py
import sys
from typing import List, Tuple

def dijkstra(e: List[List[Tuple[int, int]]], start: int, start_dist: int) -> Tuple[List[int], List[bool]]:
    n = len(e)
    added = [False] * n
    res = [float('inf')] * n
    res[start] = start_dist
    added[start] = True
    pq = [(start_dist, start)]
    while pq:
        d, i = pq.pop(0)
        if res[i] < d:
            continue
        for j, ew in e[i]:
            if not added[j] or d + ew < res[j]:
                res[j] = d + ew
                added[j] = True
                pq.append((res[j], j))
    return res, added

def solve(n: int, k: int, arrs: List[List[int]]) -> int:
    sums = [sum(arr) for arr in arrs]
    dps = [[0] * (k+1) for _ in range(13)]
    for i in range(n):
        for dpi in range(12):
            do_copy(dpi+1, dpi)
            v = sums[i]
            k1 = len(arrs[i])
            for j in range(k, k1-1, -1):
                dps[dpi][j] = max(dps[dpi][j], dps[dpi][j-k1] + v)
    return id13(lambda rec, l, r, dpi: (
        ans := 0,
        carr := arrs[l],
        dp := dps[dpi],
        if l+1 == r:
            ans = max(ans, dp[k])
            sum = 0
            for i in range(min(k, len(carr))):
                sum += carr[i]
                ans = max(ans, dp[k-i-1] + sum)
        else:
            do_split = lambda l1, r1, l2, r2: (
                do_copy(dpi+1, dpi),
                for i in range(l1, r1-1):
                    id14(i, dpi+1),
                rec(l2, r2, dpi+1)
            )
            m = (l+r) // 2
            ans = max(ans, do_split(l, m, m, r))
            ans = max(ans, do_split(m, r, l, m))
        ans
    )(0, n, 0))

def do_copy(dpi1: int, dpi2: int) -> None:
    dps[dpi1] = dps[dpi2].copy()

def id14(i: int, dpi: int) -> None:
    v = sums[i]
    k1 = len(arrs[i])
    for j in range(k, k1-1, -1):
        dps[dpi][j] = max(dps[dpi][j], dps[dpi][j-k1] + v)

def main() -> None:
    n, k = map(int, input().split())
    arrs = [list(map(int, input().split())) for _ in range(n)]
    print(solve(n, k, arrs))

if __name__ == "__main__":
    main()
